lldb/source/Plugins/ExpressionParser/Swift/SwiftUserExpression.cpp

```diff
@@ -270,6 +451,89 @@ GetPersistentState(Target *target, ExecutionContext &exe_ctx) {
   return target->GetSwiftPersistentExpressionState(*exe_scope);
 }
 
+/// Check if we can evaluate the expression as generic.
+/// Currently, evaluating expression as a generic has several limitations:
+/// - Only self will be evaluated with unbound generics.
+/// - The Self type can only have one generic parameter. 
+/// - The Self type has to be the outermost type with unbound generics.
+static bool CanEvaluateExpressionWithoutBindingGenericParams(
+    llvm::ArrayRef<SwiftASTManipulator::VariableInfo> variables, Block *block,
+    StackFrame &stack_frame) {
+  // First, find the compiler type of self with the generic parameters not
+  // bound.
+  auto self_var = SwiftExpressionParser::FindSelfVariable(block);
+  if (!self_var)
+    return false;
+
+  lldb::ValueObjectSP self_valobj =
+      stack_frame.GetValueObjectForFrameVariable(self_var,
+                                                     lldb::eNoDynamicValues);
+  if (!self_valobj)
+    return false;
+
+  auto self_type = self_valobj->GetCompilerType();
+  if (!self_type)
+    return false;
+
+  auto *ts = self_type.GetTypeSystem()
+                 .dyn_cast_or_null<TypeSystemSwift>()
+                 ->GetSwiftASTContext();
+
+  if (!ts)
+    return false;
+
+  auto swift_type = ts->GetSwiftType(self_type);
+  if (!swift_type)
+    return false;
+
+  auto *decl = swift_type->getAnyGeneric();
+  if (!decl)
+    return false;
+
+
+  auto *env = decl->getGenericEnvironment();
+  if (!env)
+    return false;
+  auto params = env->getGenericParams();
+
+  // If there aren't any generic parameters we can't evaluate the expression as
+  // generic.
+  if (params.empty())
+    return false;
+
+  auto *first_param = params[0];
+  // Currently we only support evaluating self as generic if the generic
+  // parameter is the first one.
+  if (first_param->getDepth() != 0 || first_param->getIndex() != 0)
+    return false;
+
+  llvm::SmallVector<SwiftASTManipulator::VariableInfo>
+      outermost_metadata_vars;
+  for (auto &variable : variables)
+    if (variable.IsOutermostMetadataPointer())
+      outermost_metadata_vars.push_back(variable);
+
+  // Check that all metadata belong to the outermost type, and check that we do
+  // have the metadata pointer available.
+  for (auto *pair : params) {
+    if (pair->getDepth() != 0)
+      return false;
+
+    auto var_name = "$τ_0_" + std::to_string(pair->getIndex());
```

Comment:
Slightly more efficient:
https://github.com/apple/llvm-project/blob/a52ae9f203b9f93e6693f0acc1f5faecd18ec227/lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntimeNames.cpp#L726

Reply:
> FWIW, it would be better to split this into NFC refactorings and new functionality.

Ok, I'll do some NFCs instead. First one: https://github.com/apple/llvm-project/pull/5790